Improvre pre-commit and github action

LecrisUT opened the pull request:
- Added markdown formatter. Overall I'm pleased with it, the only part that I would change is how it formats sub-items. Can you confirm the documentations are displayed correctly?
- Added pre-commit to github action. Haven't made it to automatically push a commit though
- Disable `clang-tidy` check. We need to fix those at some point
- Added test for `intel`, `gcc` and `llvm` toolchains
- Included presets to automate the runs

Closes: #243
Closes: #238

LecrisUT committed:
Pre-commit fixes

Sorry, I have disabled it locally at some point

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT committed:
Run pre-commit markdown

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT committed:
Add markdown format to pre-commit and clean-up

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT committed:
Temporarily disable `cland-tidy` pre-commit

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

atztogo commented:
Look super to me.

lan496 commented on .pre-commit-config.yaml:
```diff
@@ -1,62 +1,60 @@
-# Work around for virtualenv with python=3.10
-# https://github.com/pre-commit/pre-commit/issues/1375
-default_language_version:
-  python: python3
```
Can you leave this workaround? In my environment (python 3.10), the following error occurs without it.
```
An unexpected error has occurred: CalledProcessError: command: ('/Users/shinohara/.pyenv/versions/miniforge3-4.10/envs/spglib/bin/python3.1', '-mvirtualenv', '/Users/shinohara/.cache/pre-commit/repozdodn76_/py_env-python3.1', '-p', 'python3.1')
return code: 1
stdout:
    RuntimeError: failed to find interpreter for Builtin discover of python_spec='python3.1'
```

LecrisUT replied:
Ok, but you really need to update your Python environment. Python 3.1 has reached end-of-life in [2012](https://devguide.python.org/versions/)

The main issue here is that your `python` is symlink-ed to 3.1 while `python3` is linked to 3.10?

lan496 replied:
My environment is Python 3.10. It seems the pre-commit forgets the second digit in Python's minor version somehow (Of course there is no Python 3.1 in my environment).

LecrisUT replied:
It doesn't look like pre-commit issue, but more like a symlink issue. Can you check where the `which python` and its variants point to?

lan496 replied:
It points to `/Users/shinohara/.pyenv/versions/miniforge3-4.10/envs/spglib/bin/python`. I use miniforge on pyenv, which may be a tricky setting. When I upgrade miniforge3-4.10 to miniforge3-4.14.0-2 (latest), the error disappears.

LecrisUT replied:
🤷 No idea what's going on there, maybe if down the symbolic link chain they differ somewhere. Anyway will keep the proposed fix up for a while.

lan496 commented on .github/workflows/pypi-wheel-builds.yml:
```diff
@@ -1,10 +1,9 @@
   push:
-    branches:
-      - rc
-      - master
```
I think the `master` branch is still required to upload PyPI.

atztogo replied:
Yes, master is needed!

LecrisUT replied:
Oh. Want to change it to tags?

atztogo replied:
> Want to change it to tags?

It is also a choice. Currently master and tag are in one-to-one correspondence in my operation.

LecrisUT replied:
Mostly it's because repos tend to move away from branches named `master` and if we move to tags, it would be completely unnecessary to keep master branch up to date

atztogo replied:
`master` and `rc` are used to trigger launching to pypi and pypi-testing, respectively. So if we set-up our convention to those, we can remove both branches.

LecrisUT replied:
Can check my tests for the push tags on my fork:
https://github.com/LecrisUT/spglib/actions/runs/4241036443

atztogo replied:
About this?
```
on:
  push:
    branches:
      - rc
      - master
  pull_request:
    branches: [rc]
```

Does this
```
  pull_request:
    branches: [rc]
```
mesn a trigger when PR is made to the rc branch? It seems a good idea.

LecrisUT replied:
Indeed. I was wondering about how to allow contributors to test the github actions for pypi. But apparently if you push tags locally it also triggers the build. It is not ideal because it will also run the `upload_pypi` step, but it should not be able to push an actual release there because it wouldn't have the secrets (I confirmed it on my fork https://github.com/LecrisUT/spglib/actions/runs/4241036443/jobs/7371203802).

But yeah, maybe having it automatic on `rc` would be easier for the contributors. Maybe change the name to `test-pypi` and/or document it more in `Contributing.md`?

One concern and probably only @lan496 can check for us, are the gihub action minutes billed on the organization? I don't see any Github action billing usage for this repo on my billing report. This workflow does eat up a lot of minutes (45min/run)

atztogo replied:
> But yeah, maybe having it automatic on rc would be easier for the contributors. Maybe change the name to test-pypi and/or document it more in Contributing.md?

Yes, it would be easier to understand. Previously rc branch was used not only for pypi-testing.

atztogo replied:
Current plan is GitHub Free.

atztogo replied:
The workflow is slow because of VM emulation of different CPU architecture (arm64) for linux if I remember correctly.

lan496 replied:
At this moment (23 Feb.), Spglib organization uses 44 min out of 2000 action min/per month. So, it's almost negligible.

LecrisUT replied:
I have definitely used more than 300 minutes on this PR alone. Let's give it a few days and check again when it's updated.

lan496 replied:
Ok. I will check it again at the end of this month.

LecrisUT replied:
Oh, now I remember, the billing is only for private repos

lan496 replied:
What generous Github!

lan496 reviewed (commented):
@LecrisUT Thanks! I've checked the formatted documents look correct.

LecrisUT committed:
Move ctest to top build directory

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT committed:
Disable sanitizer on intel compilers

Maybe there is an appropriate way of introducing them, but currently it fails at link-time because it is not appropriately bundled

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT committed:
Include CMakePresets for CI

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT committed:
Add runners for all toolchains

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT commented:
Two new things:
- Added github action for the 3 main toolchains: `gcc`, `intel` and `llvm`
- Added presets to run the CI. Should be able to run them locally as well, e.g.:

```
cmake --workflow --preset intel-ci
```

Other minor things:
- Moved `enable_testing` to top-level because presets don't seem to allow running `ctest` in subfolder
- Disabled sanitizers in intel compiler. For some reason my intel package did not bundle the appropriate `asan` library that it is linked to. Maybe there is an appropriate way to fix this though
- Cannot use `flang` yet because it is lacking `c_ptr` and such

LecrisUT commented:
Yey? The github action already payed off. But I am clueless about address-sanitizers @atztogo @lan496 Can you check this run failure?

https://github.com/spglib/spglib/actions/runs/4243434293/jobs/7376187820#step:7:221

On my local Fedora machine it passes. You should be able to run it once you install the clang compiler

```
cmake --workflow --preset llvm-ci
```

Edit: Oh there are multiple libasan libraries... :thinking: 
https://github.com/spglib/spglib/actions/runs/4243434293/jobs/7376187820#step:7:157

Huh. That worked. So for some reason ubuntu package of `clang` does not link to the correct version of `libasan`. I have changed it to use Fedora and it worked as expected.

Also, sorry if you are getting bombarded by action notifications.

LecrisUT commented on .github/workflows/pypi-wheel-builds.yml:
```diff
@@ -1,36 +1,32 @@
+    tags: [ "v[0-9]+.[0-9]+.[0-9]+*" ]
   pull_request:
-    branches: [rc]
+    branches: [ test-PyPi-action ]
```
Are we ok with this, or should we keep `rc` branch instead?

One thing to note is that with the tag system, we can run the PyPi workflow even without making a PR to this branch. Not sure how much this would be used in this case.

atztogo replied:
I think we can remove `rc` and make `test-PyPi-action` branch, but how do you think @lan496?

lan496 replied:
I agree with the new branch name `test-PyPi-action`. But, the current setting seems not to call `upload_pypi` job for PR to `test-PyPi-action` branch, right?

I understand the previous branching rules were as follows:
- PR to `rc`: test to build wheels
- Push to `rc`: upload wheels to TestPyPI
- Push to `master`: upload wheels to PyPI

The current setting seems to behave as
- PR to `test-PyPi-action`: test to build wheels
- Push to `rc` with tag: upload wheels to TestPyPI
- Push to branch other than `rc` with tag: upload to PyPI

BTW, this PR becomes large now. The improvement for branching and tagging may be better to be addressed in a new PR.

LecrisUT replied:
> But, the current setting seems not to call upload_pypi job for PR to test-PyPi-action branch, right?

Specifically the uploading to pypi steps, it still uploads artifacts for local testings. And yes this is intentional because at the PR stage the branch is still under development. If we truly want to make it available we push a tag.

> The current setting seems to behave as

It is much simpler:
- Any pushed tags of the form `vX.Y.Z*` or PRs to the branch name `test-PyPi-action` will trigger a build
- If it's a tag with `rc` in it (maybe can narrow the format more can't seem to regex it though) upload to TestPyPi
- If it's a tag without `rc` or `test` in it upload to PyPi

> BTW, this PR becomes large now. The improvement for branching and tagging may be better to be addressed in a new PR.

Agree, where do you want to split it off? I have another update to fix codecov almost done as well.

lan496 replied:
@LecrisUT Thanks for your clarification. I see `github.ref` refers to the tag's name (I misread it refers to a branch name).

> Agree, where do you want to split it off? I have another update to fix codecov almost done as well.

Let's split the tags and branch parts off. The codecov fix will be proper to be included in this PR.

LecrisUT replied:
> @LecrisUT Thanks for your clarification. I see github.ref refers to the tag's name (I misread it refers to a branch name).

Actually, good point, there is a variable specific for tags that will make it clearer. But I hope there is a regex way to do it instead of Contains.

> Let's split the tags and branch parts off. 

👍

LecrisUT commented:
Welp, that was an ordeal to debug, might have eaten quite a lot of github action minutes on this, but all is up and running now, check it out.

LecrisUT committed:
Add test coverage for C and Fortran

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

atztogo commented:
> that was an ordeal to debug

Indeed. A branch for the debug may be useful, where only one pypi wheel (CPython 3.9 & manylinux x86_64). Can you use spglib organization to run the test?

LecrisUT commented:
Oh, good idea, we can edit that with environment variables and link it to the tags/MR like it is done here.

I have ran the pypi tests on my fork, and it seems to be billed on the organization.

LecrisUT committed:
Update testing github action

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT committed:
Fix python code coverage

Should only report on spglib test, not on the tests themselves.
Ideally we should only run `coverage run`, but this does not work with non-editable virtual environments

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT committed:
Clean up google tests and reworked the test labels

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT committed:
Add codecov convergence tests

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT committed:
Separate coverage reports

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT commented:
Holly batman, I've managed it. We have separate codecov coverages depending on the api:
https://app.codecov.io/gh/spglib/spglib/pull/241/flags

lan496 reviewed (approved):
@LecrisUT That's impressive! Let me get this straight roughly on how to get coverage on C APIs: The flag `--coverage` need to use gcov with gcc. Then, the coverage reports are uploaded to covecov by [lcov-action](https://github.com/danielealbano/lcov-action). Is it correct?

@atztogo I think this PR is ready to be merged. What do you think?

atztogo commented:
> @atztogo I think this PR is ready to be merged. What do you think?

Sure, I agree.

LecrisUT commented:
Indeed what we're doing here is making sure each coverage is run cleanly so we can get the correct coverage of the python api when we're calling the C functions. This came up when doing the fotran coverage and noticing that it was 0%. That's because it was calling the C api directly, so after separating the runs by labels we get the correct coverage 🎉
